### ISSUE TITLE
[Tutorial] How to broadcast a jupyterlab server with gdrive integration for remote access

### ISSUE THREAD

**Issue by PavlosMelissinos** [jupyterlab/jupyterlab-google-drive#12]
Not really an issue but I'd like to offer some help to aspiring testers/users of the project. It took me some time to actually figure out how to make this work remotely. It seems much easier in retrospect!

**Note: This is a temporary hack to circumvent #7 in a multi-user/multi-machine environment until it's resolved.**

Anyway, AFAICT, the actual problem is that google drive has to work with a localhost server in order to be able to allow the server to see your files. All is good when you're running the server locally but what happens when you need to have multiple users that connect remotely?

My solution uses ssh tunneling, if there's a better way please let me know because it's not pretty.

I assume you're behind some kind of firewalled NAT router.

# Server side
1. Follow the instructions to install jupyter lab and google drive integration and make sure everything works as expected locally.
2. First you need to port forward the jupyter lab server port over the internet, as well as the ssh port (this is really dangerous so at least don't use the default one). You should set your own values for your setup but for this demo I assume the following:
    1. Server external (public) IP address: `123.45.67.89`
    2. Jupyter port on the server: `8888`
    3. Forwarded ssh port: `22222`
    4. Forwarded jupyter port over the internet: `9888`
    5. LAN IP of the server: `192.168.1.2`
    6. Final client url: `localhost:8890`
    7. username: `user`
3. Create a user on the server for each jupyter user (they will need credentials so it's better to have their own than share yours). No need for superuser permissions.
4. Run `jupyter lab --no-browser --ip=192.168.1.2`

# Client side:

## Using putty (e.g. for windows machines with no bash access at all and/or people who are afraid of the terminal 😛)

1. Download putty
2. Setup the ssh session
![ssh session](http://i.imgur.com/6zhGrs1.png)
3. Setup the ssh tunnel (don't forget to click on add when you're finished)
![ssh session](http://i.imgur.com/aTQKKtq.png)
4. Save the configuration if you want
5. Click on "open"
6. Enter your credentials
7. Open your browser at `localhost:8890`
8. Enjoy jupyter lab with google drive integration

## Clients with direct access to ssh

1. Connect to the server via ssh to setup your user's password etc: `ssh user@123.45.67.89 -p 22222`
2. Open an ssh tunnel: `ssh -N -L localhost:8890:123.45.67.89:9888 user@123.45.67.89`
3. Open browser at `localhost:8890`
8. Enjoy jupyter lab with google drive integration

**Comment by ian-r-rose** [comment]
Thanks for this @PavlosMelissinos, this looks like a useful guide! I would just add that the restriction to `localhost` is temporary, and once #7 is resolved it will be possible to configure the extension to work with any IP (though you would need to set up your own client id in the Google Developer Console).

**Comment by PavlosMelissinos** [comment]
Oh right! Do you have a rough estimate maybe?

It's great that the project is moving so fast. In fact I've already replaced my vanilla jupyter notebooks with lab. A tabbed environment is really handy and I can't wait to have all the upcoming features!

**Comment by ian-r-rose** [comment]
It should be doable once jupyterlab/jupyterlab#2157 lands. I am not sure exactly when, but within a couple of weeks, I should think.

**Comment by blink1073** [comment]
We intend to merge that one within the next day ;).

**Comment by ian-r-rose** [comment]
Very good :)

**Comment by ian-r-rose** [comment]
Hi @PavlosMelissinos , the ability to set your own client-id (with its own set of approved origins) has been released in v0.2.0, via #13. Are you able to check that it works, and that the instructions in `advanced.md` are sufficient?

**Comment by PavlosMelissinos** [comment]
Hi, I've been busy with boring real world stuff since last week but hopefully I'll be able to check it today.

**Comment by PavlosMelissinos** [comment]
Hi again. I've followed the steps one by one but where do I need to enter the clientID? I'm not sure what the settings registry is... Also when I run jupyterlab I see the `@jupyterlab/google-drive  enabled  OK` line but there's no google drive tab on the UI.

**Comment by ian-r-rose** [comment]
For now the settings registry is only settable via the JupyterLab UI (in the file menu), and must be set client-side. We are working on having server-side settings (which would be more appropriate for this use-case), but are not there yet.

For the other issue, do you see any errors in the Javascript console?

**Comment by PavlosMelissinos** [comment]
Must be an error on my end because I get the following output:

```
Plugin 'jupyter.extensions.google-drive' failed to activate.  main.bundle.js:185973:17
	Application.prototype.start/promises</< http://localhost:8890/lab/main.bundle.js:185973:17
Error: No provider for: jupyter.services.commandpalette.
Stack trace:
Application.prototype.resolveRequiredService@http://localhost:8890/lab/main.bundle.js:185888:35
Application.prototype.activatePlugin/required<@http://localhost:8890/lab/main.bundle.js:185846:64
Application.prototype.activatePlugin@http://localhost:8890/lab/main.bundle.js:185846:24
Application.prototype.start/promises<@http://localhost:8890/lab/main.bundle.js:185972:20
Application.prototype.start@http://localhost:8890/lab/main.bundle.js:185971:24
main@http://localhost:8890/lab/main.bundle.js:131425:5
  main.bundle.js:185974:17
	Application.prototype.start/promises</<
```

Is there a way to somehow clean the installation completely without nuking the whole conda installation? This happens in a new conda environment I created especially to test this and I've already tried deleting the node_modules directory.

**Comment by ian-r-rose** [comment]
I think that this is the same error that was seen in #19. It was a bug in the extension system for JupyterLab, which has been fixed in master, but remains in 0.24. You should be able to fix it by running `jupyter lab clean && jupyter lab build`

**Comment by PavlosMelissinos** [comment]
Yup, that was it but I still don't have access to my google drive account.

I've added the json entry in the google drive editor in the settings and applied the changes.

```
{
    "clientId": <CLIENT ID VALUE>
}
```

However, if I click on the google drive tab on the side panel, I simply get the unclickable gdrive logo.
The javascript console doesn't report anything suspicious:

```
Empty string passed to getElementById().  main.bundle.js:107250:44
gapi: loaded onto page  main.bundle.js:26862:17
uncaught exception: undefined
```

I've already added the server's public IP to the Authorized JavaScript origins screen so this should work, right? Which usage scenarios should work so that I could try them out?

**Comment by ian-r-rose** [comment]
Yes, it sounds like your setup is correct.

Have you included the port number (if any) in the list of authorized Javascript origins? Also, can you verify that it works in the case of a local server? That is to say, if you add `http://localhost:8888` to the authorized origins and launch a local instance, does it work as intended?

I should be able to try out a hub install on Monday to see if I can more closely replicate your environment.

**Comment by PavlosMelissinos** [comment]
Unfortunately the server is a headless raspberry pi 3 and I currently don't
have a properly setup machine handy so I can't access the server locally.
I'll try to setup a VM tomorrow.

However, using a ssh tunnel on the raspberry pi behaves in the same way.

**Comment by PavlosMelissinos** [comment]
Also, yes I've added the port number

**Comment by PavlosMelissinos** [comment]
I just tried it again at work on a new installation, this time loading the client from the same machine as the server at localhost:8888, but I still can't log in to my google account (only the logo appears)...

On Firefox 54.0:
```
Empty string passed to getElementById().  main.bundle.js:107250:44
gapi: loaded onto page  main.bundle.js:26862:17
uncaught exception: undefined (unknown)
```

On Chrome 58.0.3029.110
```
gapi: loaded onto page
promise.js:44 Uncaught (in promise) undefined
PromiseDelegate.reject @ promise.js:44
(anonymous) @ gapi.js:100
h.o0 @ cb=gapi.loaded_0:82
xs @ cb=gapi.loaded_0:85
Wq @ cb=gapi.loaded_0:85
_.C.uea @ cb=gapi.loaded_0:84
Ap @ cb=gapi.loaded_0:78
lab:1 Uncaught (in promise) undefined
```

**Comment by ian-r-rose** [comment]
I have just tested this on a JupyterHub instance and it works on my setup. I still suspect a mismatch between the whitelisted origins and your clientId. 
Can you set a breakpoint at the point of failure (`gapi.js:100`), and inspect the `err` object? This will let us know if that is the issue.

(and clearly we need to provide a better error message in this instance than silently failing!)

**Comment by PavlosMelissinos** [comment]
Hi there, I haven't forgotten about this, it's just that the collaboration notebook has been part of a side-project which is now on hold, so it's kind of a lower priority for me at the moment, sorry.

However, my only feedback this far would be that the documentation needs to become a little bit more user-friendly and a way to achieve that is to add more screenshots. The main issue for me throughout this experience has been that I literally didn't know what I was missing 😆. Even when the gdrive functionality was missing (for instance I wouldn't see all the entries in the settings), I had no idea how the UI was supposed to look so I couldn't tell where I should add the json object.

A way to see how each change affects the UI would be great. Of course, I realize this add-on is still in early stage in the development process, so it's ok.

In any case, I doubt it's a bug and this is not really a support forum :p but I'll definitely let you know my updated thoughts when I get it to work. Until then feel free to close the issue if you want.

Thanks for the great work :)

**Comment by ian-r-rose** [comment]
Hi @PavlosMelissinos, the documentation for setting up your own client id has improved in #70 and #71, so I am marking this as resolved. If you still have issues, feel free to reopen!

**Comment by PavlosMelissinos** [comment]
Perfect 💯, thanks a lot! 😄
